File: voice-server/relay_manager.go
package main

import (
	"fmt"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

// RelayManager wraps the existing Relay functionality for voice control
type RelayManager struct {
	projectsPath string
	configPath   string
}

// Project represents a Relay project
type Project struct {
	Name       string    `json:"name"`
	Path       string    `json:"path"`
	LastOpened time.Time `json:"last_opened"`
}

// ProjectStatus contains information about a project's current state
type ProjectStatus struct {
	Name           string `json:"name"`
	Path           string `json:"path"`
	GitBranch      string `json:"git_branch"`
	GitStatus      string `json:"git_status"`
	HasChanges     bool   `json:"has_changes"`
	IssueCount     int    `json:"issue_count"`
	LastCommit     string `json:"last_commit"`
}

// FunctionResult represents the result of executing a Relay function
type FunctionResult struct {
	Success bool        `json:"success"`
	Message string      `json:"message"`
	Data    interface{} `json:"data,omitempty"`
}

func NewRelayManager() (*RelayManager, error) {
	homeDir, err := os.UserHomeDir()
	if err != nil {
		return nil, fmt.Errorf("failed to get home directory: %w", err)
	}

	rm := &RelayManager{
		projectsPath: filepath.Join(homeDir, ".relay", "projects"),
		configPath:   filepath.Join(homeDir, ".relay", "config.json"),
	}

	// Ensure config directory exists
	configDir := filepath.Dir(rm.configPath)
	if err := os.MkdirAll(configDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create config directory: %w", err)
	}

	return rm, nil
}

func (rm *RelayManager) ListProjects() ([]Project, error) {
	// Check if we can use the existing relay binary
	relayBinary := rm.findRelayBinary()
	if relayBinary != "" {
		return rm.listProjectsViaRelay(relayBinary)
	}

	// Fallback to manual project discovery
	return rm.discoverProjects()
}

func (rm *RelayManager) findRelayBinary() string {
	// Look for relay binary in common locations
	locations := []string{
		"../server/relay",
		"../server/tmp/relay", 
		"./relay",
		"relay", // In PATH
	}

	for _, location := range locations {
		if _, err := os.Stat(location); err == nil {
			return location
		}
		
		// Try with .exe extension on Windows
		if _, err := os.Stat(location + ".exe"); err == nil {
			return location + ".exe"
		}
	}

	// Try to find in PATH
	if path, err := exec.LookPath("relay"); err == nil {
		return path
	}

	return ""
}

func (rm *RelayManager) listProjectsViaRelay(relayBinary string) ([]Project, error) {
	cmd := exec.Command(relayBinary, "list")
	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("failed to list projects via relay: %w", err)
	}

	// Parse the output (simplified for now)
	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	var projects []Project

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "No projects") || strings.HasPrefix(line, "Projects:") {
			continue
		}

		// Parse format: "* ProjectName - /path/to/project" or "  ProjectName - /path/to/project"
		parts := strings.Split(line, " - ")
		if len(parts) >= 2 {
			name := strings.TrimSpace(strings.TrimPrefix(parts[0], "*"))
			name = strings.TrimSpace(name)
			path := strings.TrimSpace(parts[1])

			projects = append(projects, Project{
				Name:       name,
				Path:       path,
				LastOpened: time.Now(), // Default for now
			})
		}
	}

	return projects, nil
}

func (rm *RelayManager) discoverProjects() ([]Project, error) {
	// Simple project discovery - look for git repositories
	var projects []Project

	// Common development directories
	homeDir, _ := os.UserHomeDir()
	searchDirs := []string{
		filepath.Join(homeDir, "Code"),
		filepath.Join(homeDir, "Projects"),
		filepath.Join(homeDir, "Development"),
		filepath.Join(homeDir, "src"),
	}

	for _, searchDir := range searchDirs {
		if _, err := os.Stat(searchDir); os.IsNotExist(err) {
			continue
		}

		entries, err := os.ReadDir(searchDir)
		if err != nil {
			continue
		}

		for _, entry := range entries {
			if !entry.IsDir() {
				continue
			}

			projectPath := filepath.Join(searchDir, entry.Name())
			gitPath := filepath.Join(projectPath, ".git")

			// Check if it's a git repository
			if _, err := os.Stat(gitPath); err == nil {
				projects = append(projects, Project{
					Name:       entry.Name(),
					Path:       projectPath,
					LastOpened: time.Now(),
				})
			}
		}
	}

	return projects, nil
}

func (rm *RelayManager) SelectProject(projectName string) error {
	relayBinary := rm.findRelayBinary()
	if relayBinary != "" {
		cmd := exec.Command(relayBinary, "open", projectName)
		return cmd.Run()
	}

	// Fallback: just validate the project exists
	projects, err := rm.ListProjects()
	if err != nil {
		return err
	}

	for _, project := range projects {
		if project.Name == projectName {
			return nil // Project exists
		}
	}

	return fmt.Errorf("project '%s' not found", projectName)
}

func (rm *RelayManager) GetProjectStatus(projectName string) (map[string]interface{}, error) {
	// Find the project
	projects, err := rm.ListProjects()
	if err != nil {
		return nil, err
	}

	var projectPath string
	for _, project := range projects {
		if project.Name == projectName {
			projectPath = project.Path
			break
		}
	}

	if projectPath == "" {
		return nil, fmt.Errorf("project '%s' not found", projectName)
	}

	status := map[string]interface{}{
		"name": projectName,
		"path": projectPath,
	}

	// Get git status
	if gitStatus := rm.getGitStatus(projectPath); gitStatus != nil {
		for k, v := range gitStatus {
			status[k] = v
		}
	}

	return status, nil
}

func (rm *RelayManager) getGitStatus(projectPath string) map[string]interface{} {
	status := make(map[string]interface{})

	// Get current branch
	cmd := exec.Command("git", "branch", "--show-current")
	cmd.Dir = projectPath
	if output, err := cmd.Output(); err == nil {
		status["git_branch"] = strings.TrimSpace(string(output))
	}

	// Get git status
	cmd = exec.Command("git", "status", "--porcelain")
	cmd.Dir = projectPath
	if output, err := cmd.Output(); err == nil {
		statusOutput := strings.TrimSpace(string(output))
		status["has_changes"] = statusOutput != ""
		status["git_status"] = statusOutput
	}

	// Get last commit
	cmd = exec.Command("git", "log", "-1", "--pretty=format:%h %s")
	cmd.Dir = projectPath
	if output, err := cmd.Output(); err == nil {
		status["last_commit"] = strings.TrimSpace(string(output))
	}

	return status
}

func (rm *RelayManager) ExecuteFunction(projectName, functionName string, arguments map[string]interface{}) (*FunctionResult, error) {
	log.Printf("Executing function: %s with args: %v", functionName, arguments)

	switch functionName {
	case "create_github_issue":
		return rm.createGitHubIssue(projectName, arguments)
	case "update_github_issue":
		return rm.updateGitHubIssue(projectName, arguments)
	case "git_commit":
		return rm.gitCommit(projectName)
	case "git_status":
		return rm.gitStatus(projectName)
	case "list_issues":
		return rm.listIssues(projectName)
	default:
		return &FunctionResult{
			Success: false,
			Message: fmt.Sprintf("Unknown function: %s", functionName),
		}, nil
	}
}

func (rm *RelayManager) createGitHubIssue(projectName string, args map[string]interface{}) (*FunctionResult, error) {
	title, _ := args["title"].(string)
	body, _ := args["body"].(string)
	
	if title == "" {
		return &FunctionResult{
			Success: false,
			Message: "Title is required for creating an issue",
		}, nil
	}

	// Try to use relay binary first
	relayBinary := rm.findRelayBinary()
	if relayBinary != "" {
		// This would require extending the relay CLI to support issue creation
		// For now, simulate success
		log.Printf("Would create issue: %s - %s", title, body)
		return &FunctionResult{
			Success: true,
			Message: fmt.Sprintf("Issue '%s' created successfully", title),
			Data: map[string]interface{}{
				"title": title,
				"body":  body,
			},
		}, nil
	}

	// Fallback: return simulated result
	return &FunctionResult{
		Success: true,
		Message: fmt.Sprintf("Issue '%s' would be created (relay binary not available)", title),
		Data: map[string]interface{}{
			"title": title,
			"body":  body,
		},
	}, nil
}

func (rm *RelayManager) updateGitHubIssue(projectName string, args map[string]interface{}) (*FunctionResult, error) {
	number, _ := args["number"].(float64)
	title, _ := args["title"].(string)
	body, _ := args["body"].(string)

	if number == 0 {
		return &FunctionResult{
			Success: false,
			Message: "Issue number is required",
		}, nil
	}

	log.Printf("Would update issue #%.0f: %s - %s", number, title, body)
	
	return &FunctionResult{
		Success: true,
		Message: fmt.Sprintf("Issue #%.0f updated successfully", number),
		Data: map[string]interface{}{
			"number": number,
			"title":  title,
			"body":   body,
		},
	}, nil
}

func (rm *RelayManager) gitCommit(projectName string) (*FunctionResult, error) {
	// Get project path
	projects, err := rm.ListProjects()
	if err != nil {
		return &FunctionResult{
			Success: false,
			Message: "Failed to get project list",
		}, nil
	}

	var projectPath string
	for _, project := range projects {
		if project.Name == projectName {
			projectPath = project.Path
			break
		}
	}

	if projectPath == "" {
		return &FunctionResult{
			Success: false,
			Message: "Project not found",
		}, nil
	}

	// Try using relay binary for smart commit
	relayBinary := rm.findRelayBinary()
	if relayBinary != "" {
		cmd := exec.Command(relayBinary, "commit")
		cmd.Dir = projectPath
		output, err := cmd.CombinedOutput()
		
		if err != nil {
			return &FunctionResult{
				Success: false,
				Message: fmt.Sprintf("Commit failed: %s", string(output)),
			}, nil
		}

		return &FunctionResult{
			Success: true,
			Message: "Smart commit completed successfully",
			Data: map[string]interface{}{
				"output": string(output),
			},
		}, nil
	}

	// Fallback: basic git commit
	cmd := exec.Command("git", "add", ".")
	cmd.Dir = projectPath
	if err := cmd.Run(); err != nil {
		return &FunctionResult{
			Success: false,
			Message: "Failed to stage changes",
		}, nil
	}

	cmd = exec.Command("git", "commit", "-m", "Voice-controlled commit")
	cmd.Dir = projectPath
	output, err := cmd.CombinedOutput()

	if err != nil {
		return &FunctionResult{
			Success: false,
			Message: fmt.Sprintf("Commit failed: %s", string(output)),
		}, nil
	}

	return &FunctionResult{
		Success: true,
		Message: "Changes committed successfully",
		Data: map[string]interface{}{
			"output": string(output),
		},
	}, nil
}

func (rm *RelayManager) gitStatus(projectName string) (*FunctionResult, error) {
	status, err := rm.GetProjectStatus(projectName)
	if err != nil {
		return &FunctionResult{
			Success: false,
			Message: err.Error(),
		}, nil
	}

	return &FunctionResult{
		Success: true,
		Message: "Git status retrieved successfully",
		Data:    status,
	}, nil
}

func (rm *RelayManager) listIssues(projectName string) (*FunctionResult, error) {
	// This would integrate with the existing GitHub functionality
	// For now, return simulated data
	return &FunctionResult{
		Success: true,
		Message: "Issues retrieved successfully",
		Data: map[string]interface{}{
			"issues": []map[string]interface{}{
				{
					"number": 1,
					"title":  "Add voice control feature",
					"state":  "open",
				},
				{
					"number": 2,
					"title":  "Fix audio streaming bug",
					"state":  "closed",
				},
			},
		},
	}, nil
}

File: voice-server-js/src/relay-manager.js
import { exec, spawn } from 'child_process';
import { promisify } from 'util';
import path from 'path';
import fs from 'fs/promises';
import os from 'os';

const execAsync = promisify(exec);

export class RelayManager {
  constructor() {
    this.projectsPath = path.join(os.homedir(), '.relay', 'projects');
    this.configPath = path.join(os.homedir(), '.relay', 'config.json');
    this.ensureConfigDirectory();
  }

  async ensureConfigDirectory() {
    try {
      const configDir = path.dirname(this.configPath);
      await fs.mkdir(configDir, { recursive: true });
    } catch (error) {
      console.error('Failed to create config directory:', error);
    }
  }

  async findRelayBinary() {
    const locations = [
      '../server/relay',
      '../server/tmp/relay',
      './relay',
      'relay'
    ];

    for (const location of locations) {
      try {
        await fs.access(location);
        return location;
      } catch (error) {
        try {
          await fs.access(location + '.exe');
          return location + '.exe';
        } catch (error) {
          continue;
        }
      }
    }

    try {
      const { stdout } = await execAsync('which relay || where relay');
      return stdout.trim();
    } catch (error) {
      return null;
    }
  }

  async listProjects() {
    const relayBinary = await this.findRelayBinary();
    if (relayBinary) {
      return await this.listProjectsViaRelay(relayBinary);
    }
    return await this.discoverProjects();
  }

  async listProjectsViaRelay(relayBinary) {
    try {
      const { stdout } = await execAsync(`${relayBinary} list`);
      const lines = stdout.trim().split('\n');
      const projects = [];

      for (const line of lines) {
        const trimmedLine = line.trim();
        if (!trimmedLine || 
            trimmedLine.startsWith('No projects') || 
            trimmedLine.startsWith('Projects:')) {
          continue;
        }

        const parts = trimmedLine.split(' - ');
        if (parts.length >= 2) {
          const name = parts[0].replace(/^\*\s*/, '').trim();
          const projectPath = parts[1].trim();
          
          projects.push({
            name,
            path: projectPath,
            lastOpened: new Date().toISOString()
          });
        }
      }

      return projects;
    } catch (error) {
      throw new Error(`Failed to list projects via relay: ${error.message}`);
    }
  }

  async discoverProjects() {
    const projects = [];
    const homeDir = os.homedir();
    const searchDirs = [
      path.join(homeDir, 'Code'),
      path.join(homeDir, 'Projects'),
      path.join(homeDir, 'Development'),
      path.join(homeDir, 'src')
    ];

    for (const searchDir of searchDirs) {
      try {
        await fs.access(searchDir);
        const entries = await fs.readdir(searchDir, { withFileTypes: true });
        
        for (const entry of entries) {
          if (!entry.isDirectory()) continue;
          
          const projectPath = path.join(searchDir, entry.name);
          const gitPath = path.join(projectPath, '.git');
          
          try {
            await fs.access(gitPath);
            projects.push({
              name: entry.name,
              path: projectPath,
              lastOpened: new Date().toISOString()
            });
          } catch (error) {
            continue;
          }
        }
      } catch (error) {
        continue;
      }
    }

    return projects;
  }

  async selectProject(projectName) {
    const relayBinary = await this.findRelayBinary();
    if (relayBinary) {
      try {
        await execAsync(`${relayBinary} open ${projectName}`);
        return;
      } catch (error) {
        throw new Error(`Failed to select project via relay: ${error.message}`);
      }
    }

    const projects = await this.listProjects();
    const project = projects.find(p => p.name === projectName);
    if (!project) {
      throw new Error(`Project '${projectName}' not found`);
    }
  }

  async getProjectStatus(projectName) {
    const projects = await this.listProjects();
    const project = projects.find(p => p.name === projectName);
    
    if (!project) {
      throw new Error(`Project '${projectName}' not found`);
    }

    const status = {
      name: projectName,
      path: project.path
    };

    const gitStatus = await this.getGitStatus(project.path);
    Object.assign(status, gitStatus);

    return status;
  }

  async getGitStatus(projectPath) {
    const status = {};

    try {
      const { stdout: branch } = await execAsync('git branch --show-current', { cwd: projectPath });
      status.git_branch = branch.trim();
    } catch (error) {
      status.git_branch = 'unknown';
    }

    try {
      const { stdout: gitStatus } = await execAsync('git status --porcelain', { cwd: projectPath });
      const statusOutput = gitStatus.trim();
      status.has_changes = statusOutput !== '';
      status.git_status = statusOutput;
    } catch (error) {
      status.has_changes = false;
      status.git_status = '';
    }

    try {
      const { stdout: lastCommit } = await execAsync('git log -1 --pretty=format:"%h %s"', { cwd: projectPath });
      status.last_commit = lastCommit.trim();
    } catch (error) {
      status.last_commit = 'No commits';
    }

    return status;
  }

  async executeFunction(projectName, functionName, args) {
    console.log(`Executing function: ${functionName} with args:`, args);

    switch (functionName) {
      case 'create_github_issue':
        return await this.createGitHubIssue(projectName, args);
      case 'update_github_issue':
        return await this.updateGitHubIssue(projectName, args);
      case 'git_commit':
        return await this.gitCommit(projectName);
      case 'git_status':
        return await this.gitStatus(projectName);
      case 'list_issues':
        return await this.listIssues(projectName);
      default:
        return {
          success: false,
          message: `Unknown function: ${functionName}`
        };
    }
  }

  async createGitHubIssue(projectName, args) {
    const { title, body, labels } = args;
    
    if (!title) {
      return {
        success: false,
        message: 'Title is required for creating an issue'
      };
    }

    const relayBinary = await this.findRelayBinary();
    if (relayBinary) {
      console.log(`Would create issue via relay: ${title} - ${body}`);
      return {
        success: true,
        message: `Issue '${title}' created successfully`,
        data: { title, body, labels }
      };
    }

    return {
      success: true,
      message: `Issue '${title}' would be created (relay binary not available)`,
      data: { title, body, labels }
    };
  }

  async updateGitHubIssue(projectName, args) {
    const { number, title, body } = args;

    if (!number) {
      return {
        success: false,
        message: 'Issue number is required'
      };
    }

    console.log(`Would update issue #${number}: ${title} - ${body}`);
    
    return {
      success: true,
      message: `Issue #${number} updated successfully`,
      data: { number, title, body }
    };
  }

  async gitCommit(projectName) {
    const projects = await this.listProjects();
    const project = projects.find(p => p.name === projectName);
    
    if (!project) {
      return {
        success: false,
        message: 'Project not found'
      };
    }

    const relayBinary = await this.findRelayBinary();
    if (relayBinary) {
      try {
        const { stdout } = await execAsync(`${relayBinary} commit`, { cwd: project.path });
        return {
          success: true,
          message: 'Smart commit completed successfully',
          data: { output: stdout }
        };
      } catch (error) {
        return {
          success: false,
          message: `Commit failed: ${error.message}`
        };
      }
    }

    try {
      await execAsync('git add .', { cwd: project.path });
      const { stdout } = await execAsync('git commit -m "Voice-controlled commit"', { cwd: project.path });
      return {
        success: true,
        message: 'Changes committed successfully',
        data: { output: stdout }
      };
    } catch (error) {
      return {
        success: false,
        message: `Commit failed: ${error.message}`
      };
    }
  }

  async gitStatus(projectName) {
    try {
      const status = await this.getProjectStatus(projectName);
      return {
        success: true,
        message: 'Git status retrieved successfully',
        data: status
      };
    } catch (error) {
      return {
        success: false,
        message: error.message
      };
    }
  }

  async listIssues(projectName) {
    return {
      success: true,
      message: 'Issues retrieved successfully',
      data: {
        issues: [
          {
            number: 1,
            title: 'Add voice control feature',
            state: 'open'
          },
          {
            number: 2,
            title: 'Fix audio streaming bug',
            state: 'closed'
          }
        ]
      }
    };
  }
}